src/irc.py

```diff
@@ -0,0 +1,181 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Bind charm business logic."""
+
+import logging
+import os
+import pathlib
+import shutil
+import tempfile
+import time
+import typing
+
+from charms.operator_libs_linux.v2 import snap
+
+import constants
+import exceptions
+
+logger = logging.getLogger(__name__)
+
+
+class ReloadError(exceptions.SnapError):
+    """Exception raised when unable to reload the service."""
+
+
+class StartError(exceptions.SnapError):
+    """Exception raised when unable to start the service."""
+
+
+class StopError(exceptions.SnapError):
+    """Exception raised when unable to stop the service."""
+
+
+class InstallError(exceptions.SnapError):
+    """Exception raised when unable to install dependencies for the service."""
+
+
+class IRCBRidgeService:
+    """IRC Bridge service class.
+
+    This class provides the necessary methods to manage the matrix-appservice-irc service.
+    The service requires a connection to a (PostgreSQL) database and to a Matrix homeserver.
+    Both of these will be part of the configuration file created by this class.
+    Once the configuration file is created, a PEM file will be generated and an app registration file.
+    The app registration file will be used to register the bridge with the Matrix homeserver.
+    PEM and the configuration file will be used by the matrix-appservice-irc service.
+    """
+
+    def reload(self) -> None:
+        """Reload the matrix-appservice-irc service.
+
+        Raises:
+            ReloadError: when encountering a SnapError
+        """
+        try:
+            cache = snap.SnapCache()
+            charmed_bind = cache[constants.IRC_BRIDGE_SNAP_NAME]
+            charmed_bind.restart(reload=True)
+        except snap.SnapError as e:
+            error_msg = (
+                f"An exception occurred when reloading {constants.IRC_BRIDGE_SNAP_NAME}. Reason: {e}"
+            )
+            logger.error(error_msg)
```

Comment:
I think `logger.exception` might be enough here :D

Reply:
and for other exception errors :) https://docs.python.org/3/library/logging.html#logging.Logger.exception

Reply:
Ah I see, so that's like `logger.error` with the additional benefit that it adds the Exception info, cool! Thanks.